### PR TITLE
[Bug fixes] Add default arg to enhance varbase ClearGradient func

### PR DESCRIPTION
**Pull request by veyron95**
### PR types
Bug fixes 

### PR changes
APIs 

#### 1. Add arg to enhance varbase ClearGradient func,
use case: 1
```
 linear.weight.clear_gradient() #  same as the original version.
```
use case: 2
```
 linear.weight.clear_gradient(False)   # False means to real clear gradient
 # default args: set_to_zero = True
```

#### 2. Expose two func to set and get the " is_empty_ " attribute of grad_var_
- _is_gradient_set_empty()
- _gradient_set_empty(is_empty = False) or _gradient_set_empty(is_empty = True) , default=True
```
@dygraph_guard
    def test_tensor_method_clear_gradient_case2(self):
        input = paddle.uniform(self.input_shape)
        linear = paddle.nn.Linear(2, 3)
        out = linear(input)
        out.backward()
        # default arg set_to_zero is true
        # so, False means real clear gradient
        linear.weight.clear_gradient(False)

        # before ._gradient_set_empty(False), 
        # the return of ._is_gradient_set_empty() should be True
        self.assertTrue(linear.weight._is_gradient_set_empty())

        # reset, because ClearGradient will call SetIsEmpty(True), but this is not our expectation.
        linear.weight._gradient_set_empty(False)
        # after ._gradient_set_empty(False), 
        # the return of ._is_gradient_set_empty() should be False
        self.assertFalse(linear.weight._is_gradient_set_empty())

        # actual result
        gradient_actual = linear.weight.grad
        # expected result
        self.assertTrue(np.empty(gradient_actual))

```

### CODE REVIEW

**Review by Baibaifan — DISMISSED** [review]
LGTM for collective

File: paddle/fluid/imperative/layer.cc [hunk 1]
```diff
@@ -219,6 +222,23 @@ void VarBase::ClearGradient() {
   }
 }
 
+void VarBase::_GradientSetEmpty(bool set_is_empty) {
```
Comment (JiabinYang) [inline paddle/fluid/imperative/layer.cc]: `set_is_empty` -> `is_empty`

File: paddle/fluid/imperative/layer.cc [hunk 2]
```diff
@@ -219,6 +222,23 @@ void VarBase::ClearGradient() {
...
+  VLOG(4) << "Gradient " << Name() << " SetIsEmpty " << set_is_empty;
```
Comment (JiabinYang) [inline paddle/fluid/imperative/layer.cc]: same

File: paddle/fluid/imperative/layer.cc [hunk 3]
```diff
@@ -219,6 +222,23 @@ void VarBase::ClearGradient() {
...
+  if (grad_var_) {
+    grad_var_->SharedVar()->SetIsEmpty(set_is_empty);
+  }
+}
+
+bool VarBase::_IsGradientSetEmpty() {
+  bool res = true;
+  if (grad_var_) {
+    res = grad_var_->SharedVar()->is_empty_;
```
Comment (JiabinYang) [inline paddle/fluid/imperative/layer.cc]: not safe an access of SharedVar, check if it's nullptr first

**Review by JiabinYang — COMMENTED** [review]
some comments

File: paddle/fluid/imperative/layer.cc (hunk 1 above)
Comment (veyron95) [reply paddle/fluid/imperative/layer.cc]: Done.

File: paddle/fluid/imperative/layer.cc (hunk 2 above)
Comment (veyron95) [reply paddle/fluid/imperative/layer.cc]: Done.

File: paddle/fluid/imperative/layer.cc (hunk 3 above)
Comment (veyron95) [reply paddle/fluid/imperative/layer.cc]: Done.

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request merged**